### PR TITLE
gh-46771: docs for asyncio.timeout

### PR DESCRIPTION
**Pull request by Tinche**
Add docs for `asyncio.timeout`, `asyncio.timeout_at` and a short section on asyncio cancellation.

* Issue: gh-46771

### CODE REVIEW

File: Doc/library/asyncio-task.rst [hunk 1]
```diff
@@ -282,6 +282,23 @@ Creating Tasks
...
 
 
+Task Cancellation
+=================
+
+Tasks can easily and safely be cancelled.
+When a task is cancelled, asyncio will raise a :exc:`asyncio.CancelledError`
+into the task at the next opportunity.
+
+Coroutines are recommended to use ``try/finally`` blocks to robustly
```
Comment (gvanrossum) [inline Doc/library/asyncio-task.rst:292]: The grammar here gets my hackles up a bit but I'm at a loss for how to put it better. (@CAM-Gerlach ?)

File: Doc/library/asyncio-task.rst [hunk 2]
```diff
@@ -282,6 +282,23 @@ Creating Tasks
...
+perform clean-up logic. In the case the :exc:`asyncio.CancelledError`
+is explicitly caught, it should generally be propagated when
+clean-up is complete. Most code can safely ignore :exc:`asyncio.CancelledError`.
+
+Important asyncio components, like :class:`asyncio.TaskGroup` and the
+:func:`asyncio.timeout` context manager, are implemented using cancellation
+internally.
```
Comment (gvanrossum) [inline Doc/library/asyncio-task.rst:299]: Maybe add "and might misbehave if a couroutine swallows `CancelledError`"?

File: Doc/library/asyncio-task.rst [hunk 3]
```diff
@@ -537,6 +554,90 @@ Shielding From Cancellation
...
 
+.. coroutinefunction:: timeout(delay)
+
+    A convenient way to limit the amount of time spent waiting on
+    something is to use the :func:`asyncio.timeout`
+    :ref:`asynchronous context manager <async-context-managers>`.
+
+    *delay* can either be ``None`` or a float or int number of
+    seconds to wait. If *delay* is ``None``, no time limit will
+    be applied. In either case, the context manager can be
```
Comment (gvanrossum) [inline Doc/library/asyncio-task.rst:565]: Maybe explain why supporting `delay=None` is useful.

File: Doc/library/asyncio-task.rst [hunk 4]
```diff
@@ -537,6 +554,90 @@ Shielding From Cancellation
...
+    rescheduled after creation.
```
Comment (gvanrossum) [inline Doc/library/asyncio-task.rst:566]: Explain how to reschedule.

File: Doc/library/asyncio-task.rst [hunk 5]
```diff
@@ -537,6 +554,90 @@ Shielding From Cancellation
...
+            try:
+                async with asyncio.timeout(10):
+                    await long_running_task()
+            except TimeoutError:
+                print("The long operation timed out, but we've handled it.")
+
+            print("This statement will run regardless.")
+
+    The context manager produced by :func:`asyncio.timeout` can be
+    rescheduled to a different deadline and inspected.
```
Comment (gvanrossum) [inline Doc/library/asyncio-task.rst:598]: This should use a new function heading so it's clear that this is about `reschedule` (and that gets added to the index). You may actually have to change the structure of the whole section to admit the existence of the `Timeout` class so you can document its methods (aren't there a few more worth documenting?)

**Review by gvanrossum — COMMENTED** [review]
Do you want to write a blurb? I'm not sure if docs-only patches generally skip this or not (@CAM-Gerlach do you know?)

**Comment by gvanrossum** [comment]
Oh, one more thing. Your PR needs to go to the main branch, we will then mark it for automatic backport to 3.11.

File: Doc/library/asyncio-task.rst (hunk 1 above)
Comment (CAM-Gerlach) [reply Doc/library/asyncio-task.rst:292]: Yeah, not only that but it is also potentially ambiguous (at least to a relative asyncio n00b like myself), as "coroutines" can be parsed either as the object of the sentence (i.e., that they are the target of the recommendation, which I'm guessing is the intended meaning), or that they are recommended _in order to use_ ``try/finally`` blocks.

Assuming the former, what about just

```suggestion
It is recommended that coroutines use ``try/finally`` blocks to robustly
```

Sure, it uses passive voice, but IMO its a good deal clearer

File: Doc/library/asyncio-task.rst [hunk 6]
```diff
@@ -282,6 +282,23 @@ Creating Tasks
       Added the *context* parameter.
 
 
+Task Cancellation
+=================
+
+Tasks can easily and safely be cancelled.
+When a task is cancelled, asyncio will raise a :exc:`asyncio.CancelledError`
+into the task at the next opportunity.
```
Comment (CAM-Gerlach) [inline Doc/library/asyncio-task.rst:290]: I may just be a n00b, but what exactly does it mean to raise an error _into_ a task? What determines when there will be an "opportunity"?

File: Doc/library/asyncio-task.rst [hunk 7]
```diff
@@ -537,6 +554,90 @@ Shielding From Cancellation
 Timeouts
 ========
 
+.. coroutinefunction:: timeout(delay)
+
+    A convenient way to limit the amount of time spent waiting on
+    something is to use the :func:`asyncio.timeout`
+    :ref:`asynchronous context manager <async-context-managers>`.
+
+    *delay* can either be ``None`` or a float or int number of
```
Comment (CAM-Gerlach) [inline Doc/library/asyncio-task.rst:563]:
```suggestion
    *delay* can either be ``None``, or a float/int number of
```

Avoid awkward and somewhat difficult to parse grammar

File: Doc/library/asyncio-task.rst [hunk 8]
```diff
@@ -537,6 +554,90 @@ Shielding From Cancellation
...
+    Example::
+
+        async def main():
+            async with asyncio.timeout(10):
+                await long_running_task()
+
+    If ``long_running_task`` takes more than 10 seconds to complete,
+    the context manager will cancel the current task and handle
+    the resulting :exc:`asyncio.CancelledError` internally, transforming it
+    into a :exc:`asyncio.TimeoutError` which can be caught and handled.
```
Comment (CAM-Gerlach) [inline Doc/library/asyncio-task.rst:577]:
```suggestion
    into an :exc:`asyncio.TimeoutError` which can be caught and handled.
```

Typo

**Review by CAM-Gerlach — COMMENTED** [review]
A few minor textual suggestions and one n00b question

**Comment by CAM-Gerlach** [comment]
> Do you want to write a blurb? I'm not sure if docs-only patches generally skip this or not (@CAM-Gerlach do you know?)

I've generally been not adding them unless the change involves documenting previously undocumented functionality, adding deprecation/discouragement messages, was very large or was otherwise meaningful for end users. I'd just been going by the devguide, which [states](https://devguide.python.org/core-developers/committing/?highlight=news#updating-news-and-what-s-new-in-python):

> Changes that affect only documentation generally do not require a NEWS entry. 

I'd been wondering myself if actual convention or current thinking differs from this—but considering _you_ are asking _me_, I figure no one really knows, heh...

**Comment by gvanrossum** [comment]
I'd use something like git cherry-pick to move your commits to a new branch created off main, and create a new PR from that. This one's dead. :-)

**Pull request closed without merging**